File: tdp_system/output_stock_web.py
"""On-screen shortage review using the same atomic remap safeguards as Excel."""
from datetime import date
from decimal import Decimal, InvalidOperation
import sqlite3

from flask import jsonify, request

try:
    from .output_stock_remap import RemapError, _snapshot, _source_hash, _store, _evaluate, _editable_period
    from .stock_tax_policy import kkknt_codes
except ImportError:
    from output_stock_remap import RemapError, _snapshot, _source_hash, _store, _evaluate, _editable_period
    from stock_tax_policy import kkknt_codes


def snapshot_for_web(conn, start, end):
    try:
        if date.fromisoformat(start) > date.fromisoformat(end):
            raise ValueError()
    except (TypeError, ValueError):
        raise RemapError('Chọn Từ ngày và Đến ngày hợp lệ, theo đúng thứ tự.')
    if not conn.in_transaction:
        conn.execute('BEGIN')
    return _snapshot(conn, start, end)


def review_shortages(conn, start, end):
    snapshot = snapshot_for_web(conn, start, end)
    exempt = kkknt_codes(conn)
    stocks = {r['product_code']: r for r in snapshot['stock']}
    blocked_reason = ''
    try:
        _editable_period(conn, start)
    except RemapError as exc:
        blocked_reason = str(exc)
    items = []
    for stock in snapshot['stock']:
        if stock['closing_qty'] >= -1e-9:
            continue
        code = stock['product_code']
        sources = [dict(ledger_id=r['id'], invoice=r['cells'][2], date=r['txn_date'],
                        source_name=r['source_item_name'], qty=-r['qty_delta'])
                   for r in snapshot['rows'] if r['product_code'] == code]
        items.append(dict(stock, sources=sources, supplementary_allowed=code in exempt))
    items.sort(key=lambda item: (item['supplementary_allowed'], item['product_code']))
    catalog = [dict(r, closing_qty=stocks.get(r['code'], {}).get('closing_qty', 0))
               for r in snapshot['catalog']]
    return dict(from_date=start, to_date=end, version=_source_hash(snapshot),
                items=items, catalog=catalog, blocked_reason=blocked_reason)


def preview_web(conn, body):
    snapshot = snapshot_for_web(conn, body.get('from'), body.get('to'))
    issues = []

    def issue(field, message):
        issues.append(dict(field=field, message=message))

    if body.get('version') != _source_hash(snapshot):
        issue('reload', 'Số liệu kho vừa thay đổi. Bấm Cập nhật số liệu rồi xem trước lại; mã hàng và số lượng đang nhập được giữ lại.')
        return dict(can_confirm=False, issues=issues, changes=[])
    try:
        _editable_period(conn, snapshot['from'])
    except RemapError as exc:
        issue('period', str(exc))
        return dict(can_confirm=False, issues=issues, changes=[])
    code = str(body.get('product_code') or '')
    stock = next((r for r in snapshot['stock'] if r['product_code'] == code), None)
    source = next((r for r in snapshot['rows'] if r['id'] == body.get('ledger_id') and r['product_code'] == code), None)
    target = next((r for r in snapshot['catalog'] if r['code'] == body.get('new_code')), None)
    if not stock or stock['closing_qty'] >= -1e-9:
        issue('reload', 'Mặt hàng này hiện không còn âm. Bấm Cập nhật số liệu để xem lại.')
    if code in kkknt_codes(conn):
        issue('supplement', 'Hàng KKKNT: dùng Lập bảng kê bổ sung khi có hàng thực mua chưa ghi nhận.')
    if not source:
        issue('ledger_id', 'Chọn Dòng xuất cần sửa trong danh sách của mặt hàng này.')
    if not target or target['code'] == code:
        issue('new_code', 'Chọn Mặt hàng thực tế đã xuất khác mã đang bị trừ nhầm, từ danh mục bên dưới.')
    elif source and str(target['unit']).strip().casefold() != str(source['unit']).strip().casefold():
        issue('new_code', 'Mặt hàng thực tế đã xuất phải cùng đơn vị tính với dòng đang sửa. Không tự đổi số lượng giữa hai đơn vị.')
    try:
        qty = Decimal(str(body.get('qty')))
        if not qty.is_finite() or qty <= 0:
            raise ValueError()
        limit = min(Decimal(str(-stock['closing_qty'])), Decimal(str(-source['qty_delta']))) if stock and source else Decimal(0)
        if qty > limit + Decimal('0.000000001'):
            issue('qty', f'Số lượng sửa tối đa {float(max(limit, 0)):g}: không vượt phần còn âm hoặc lượng của dòng xuất đã chọn.')
    except (InvalidOperation, ValueError, TypeError):
        qty = Decimal(0)
        issue('qty', 'Nhập Số lượng sửa là số lớn hơn 0.')
    if issues:
        return dict(can_confirm=False, issues=issues, changes=[])
    change = dict(ledger_id=source['id'], line_id=source['source_line_id'],
                  invoice_id=source['source_invoice_id'], old_code=code, old_name=source['name'],
                  new_code=target['code'], new_name=target['name'], qty=float(qty),
                  old_unit=source['unit'], unit=target['unit'], source_stock_qty=-source['qty_delta'],
                  old_unit_cost=source['unit_cost'], excel_row=1)
    errors = _evaluate(conn, snapshot, [change])
    # Do not offer another negative item as a convenient recipient, including KKKNT.
    if change['new_closing_after'] < -1e-9:
        errors = ['Mặt hàng thực tế đã xuất không đủ tồn để nhận số lượng này. Kiểm tra hàng nhập còn thiếu hoặc sửa Số lượng sửa theo thực tế.']
    for error in errors:
        issue('new_code', error.replace('hàng Excel', 'dòng xuất').replace('Excel', '').replace('Mã nhận', 'Mặt hàng thực tế đã xuất'))
    token = None
    if not issues:
        token = _store(conn, 'preview', dict(snapshot=snapshot, changes=[change], errors=[],
                                           mode='deficit_only_v1', origin='web'))
    return dict(token=token, can_confirm=not issues, issues=issues, changes=[change])


def register_routes(app, ctx):
    @app.get('/api/inventory/output-remap/shortages')
    def output_web_shortages():
        try:
            with ctx['db']() as conn:
                result = review_shortages(conn, request.args.get('from'), request.args.get('to'))
            return jsonify(ok=True, **result)
        except ValueError as exc:
            return jsonify(ok=False, error=str(exc)), 400
        except sqlite3.Error:
            return jsonify(ok=False, error='Kho đang được cập nhật. Bấm Cập nhật số liệu để thử lại.'), 409

    @app.post('/api/inventory/output-remap/web-preview')
    def output_web_preview():
        try:
            with ctx['db']() as conn:
                result = preview_web(conn, request.get_json(silent=True) or {})
            return jsonify(ok=True, **result)
        except ValueError as exc:
            return jsonify(ok=False, error=str(exc)), 400
        except sqlite3.Error:
            return jsonify(ok=False, error='Chưa xem trước được vì kho đang được cập nhật. Thông tin đã nhập vẫn còn; bấm Cập nhật số liệu rồi xem trước lại.'), 409

File: tdp_system/static/output-stock-web.js
/* Direct correction of stock identities; posting requires a reviewed confirmation. */
window.TdpOutputStockWeb = function (options) {
  const {api, esc, from, to} = options;
  const dialog = document.createElement('dialog');
  dialog.className = 'inventory-totals-dialog output-stock-web';
  dialog.setAttribute('aria-label', 'Xử lý hàng âm');
  dialog.innerHTML = '<div class="inventory-totals-heading"><h3>Xử lý hàng âm</h3><button class="btn btn-outline" data-close>Đóng</button></div>' +
    '<p>Kỳ ' + esc(from) + ' → ' + esc(to) + '. Chọn đúng mặt hàng để xem nguyên nhân và xử lý ngay tại đây.</p>' +
    '<div class="form-actions"><button class="btn btn-outline" data-reload>Cập nhật số liệu</button><button class="btn btn-outline" data-excel>Cách khác: sửa bằng Excel</button></div>' +
    '<div data-status role="status"></div><div data-issues role="alert"></div>' +
    '<div data-list></div><div data-editor></div><div data-preview></div>';
  document.body.appendChild(dialog); dialog.showModal();
  let data, activeCode = options.code || '', busy = false, preview = null;
  const el = selector => dialog.querySelector(selector);
  const qty = value => Number(value || 0).toLocaleString('vi-VN', {maximumFractionDigits:6});
  const field = name => el('[name="' + name + '"]');
  const post = body => ({method:'POST', headers:{'Content-Type':'application/json'}, body:JSON.stringify(body)});
  function focus(name) {
    if (name === 'dates') { dialog.close(); options.onDates(); return; }
    if (name === 'reload') { el('[data-reload]').focus(); return; }
    if (name === 'period') { el('[data-period]')?.click(); return; }
    if (name === 'supplement') { el('[data-supplement]')?.click(); return; }
    const target = field(name); if (target) { target.scrollIntoView({block:'center'}); target.focus(); }
  }
  function issues(items) {
    el('[data-issues]').innerHTML = items.map((item, i) => '<p class="error-summary">' + esc(item.message) +
      ' <button type="button" class="btn btn-outline" data-fix="' + i + '">' +
      esc(item.field === 'reload' ? 'Cập nhật số liệu' : item.field === 'dates' ? 'Sửa khoảng ngày' : item.field === 'period' ? 'Mở kỳ kho' : 'Sửa tại đây') + '</button></p>').join('');
    el('[data-issues]').querySelectorAll('[data-fix]').forEach(button => button.onclick = () => {
      const target = items[Number(button.dataset.fix)].field;
      if (target === 'reload') load(true); else focus(target);
    });
    if (items.length) el('[data-issues]').scrollIntoView({block:'nearest'});
  }
  function working(value) {
    busy = value;
    dialog.querySelectorAll('button,input,select').forEach(control => { control.disabled = value; });
    if (!value && el('[data-confirm]')) el('[data-confirm]').disabled = !field('confirmed').checked;
  }
  function invalidate() { preview = null; el('[data-preview]').innerHTML = ''; issues([]); }
  function renderList() {
    el('[data-list]').innerHTML = '<div class="table-wrap"><table><thead><tr><th>Mã / tên hàng</th><th>Tồn cuối kỳ</th><th>Xử lý</th></tr></thead><tbody>' +
      data.items.map(item => '<tr><td>' + esc(item.product_code + ' · ' + item.product_name) + '</td><td>' + qty(item.closing_qty) + ' ' + esc(item.unit) + '</td><td><button class="btn btn-primary" data-item="' + esc(item.product_code) + '">Xem nguyên nhân và xử lý</button></td></tr>').join('') +
      (!data.items.length ? '<tr><td colspan="3">Không có hàng âm trong kỳ đã chọn.</td></tr>' : '') + '</tbody></table></div>';
    el('[data-list]').querySelectorAll('[data-item]').forEach(button => button.onclick = () => {
      activeCode = button.dataset.item; renderEditor();
    });
  }
  function renderEditor(saved) {
    invalidate();
    const item = data.items.find(row => row.product_code === activeCode);
    if (!item) { el('[data-editor]').innerHTML = activeCode ? '<p>' + esc(activeCode) + ': không còn âm trong kỳ đã chọn.</p>' : ''; return; }
    const unit = esc(item.unit);
    el('[data-editor]').innerHTML = '<section class="card"><h3>' + esc(item.product_code + ' · ' + item.product_name) + '</h3>' +
      '<div class="stock-cause-totals">' + [['Tồn đầu kỳ',item.opening_qty],['Nhập trong kỳ',item.input_qty],['Xuất trong kỳ',item.output_qty],['Tồn cuối kỳ',item.closing_qty]].map(pair => '<div><span>' + pair[0] + '</span><strong>' + qty(pair[1]) + ' ' + unit + '</strong></div>').join('') + '</div>' +
      (item.reversal_qty ? '<p>Hoàn tác trong kỳ: ' + qty(item.reversal_qty) + ' ' + unit + '.</p>' : '') +
      '<p>Cần đối chiếu ' + qty(-item.closing_qty) + ' ' + unit + '. Nếu hàng thực mua chưa ghi kho, kiểm tra đầu vào. Nếu dòng xuất đang trừ nhầm mã, chọn đúng mặt hàng thực tế đã xuất bên dưới.</p>' +
      '<div class="form-actions"><button class="btn btn-outline" data-input>Kiểm tra hóa đơn đầu vào</button>' +
      (item.supplementary_allowed ? '<button class="btn btn-primary" data-supplement>Lập bảng kê bổ sung cho mặt hàng này</button>' : '') + '</div>' +
      (data.blocked_reason ? '<p class="error-summary">' + esc(data.blocked_reason) + '</p><button class="btn btn-outline" data-period>Mở kỳ kho</button>' : '') +
      (!data.blocked_reason && !item.supplementary_allowed && item.sources.length ? '<form data-edit novalidate><h4>Sửa mã của dòng xuất bị trừ nhầm</h4>' +
        '<label>Dòng xuất cần sửa<select name="ledger_id">' + item.sources.map(row => '<option value="' + row.ledger_id + '">' + esc(row.date + ' · HĐ ' + row.invoice + ' · ' + row.source_name + ' · ' + qty(row.qty) + ' ' + item.unit) + '</option>').join('') + '</select></label>' +
        '<label>Mặt hàng thực tế đã xuất<input name="new_code" list="stock-web-products" placeholder="Gõ mã hoặc tên để chọn trong danh mục" autocomplete="off"></label>' +
        '<datalist id="stock-web-products">' + data.catalog.filter(p => p.code !== item.product_code && p.unit.trim().toLocaleLowerCase() === item.unit.trim().toLocaleLowerCase()).map(p => '<option value="' + esc(p.code) + '">' + esc(p.name + ' · Tồn cuối kỳ: ' + qty(p.closing_qty) + ' ' + p.unit) + '</option>').join('') + '</datalist><p data-target></p>' +
        '<label>Số lượng sửa (' + unit + ')<input name="qty" type="number" step="any" min="0" value="' + Math.min(-item.closing_qty, item.sources[0].qty) + '"></label>' +
        '<p>Chỉ sửa phần bị trừ nhầm, tối đa bằng phần còn âm và lượng của dòng xuất đã chọn. Mặt hàng nhận phải đúng hàng thực tế đã xuất, cùng đơn vị tính.</p><button class="btn btn-primary" type="submit">Xem trước thay đổi tồn kho</button></form>' :
        (!item.supplementary_allowed && !data.blocked_reason ? '<p>Không có dòng hóa đơn đã ký, đã ghi kho có thể sửa mã trong kỳ này. Kiểm tra đầu vào hoặc đối chiếu tồn đầu kỳ; không tự cộng kho để xóa âm.</p>' : '')) + '</section>';
    el('[data-input]').onclick = async () => {
      working(true);
      try { await options.onInput(activeCode); dialog.close(); }
      catch (error) { issues([{field:'reload', message:error.message}]); }
      finally { working(false); }
    };
    if (el('[data-supplement]')) el('[data-supplement]').onclick = () => options.onSupplement(activeCode, () => load(true));
    if (el('[data-period]')) el('[data-period]').onclick = () => options.onPeriod();
    if (el('[data-edit]')) {
      if (saved) ['ledger_id','new_code','qty'].forEach(key => { if (saved[key] != null) field(key).value = saved[key]; });
      const targetInfo = () => {
        const target = data.catalog.find(p => p.code === field('new_code').value.trim());
        el('[data-target]').textContent = target ? target.name + ' · Tồn cuối kỳ: ' + qty(target.closing_qty) + ' ' + target.unit : '';
      };
      targetInfo(); field('new_code').addEventListener('input', targetInfo);
      field('ledger_id').onchange = () => {
        const source = item.sources.find(r => String(r.ledger_id) === field('ledger_id').value);
        if (source) field('qty').value = Math.min(-item.closing_qty, source.qty);
        invalidate();
      };
      el('[data-edit]').onsubmit = check;
      requestAnimationFrame(() => focus(saved ? 'new_code' : 'ledger_id'));
    } else el('[data-editor]').scrollIntoView({block:'nearest'});
  }
  function values() {
    return field('ledger_id') ? {ledger_id:Number(field('ledger_id').value), new_code:field('new_code').value.trim(), qty:field('qty').value} : null;
  }
  async function check(event) {
    event.preventDefault(); if (busy) return;
    invalidate(); working(true); el('[data-status]').textContent = 'Đang kiểm tra thay đổi. Kho chưa được ghi nhận…';
    try {
      preview = await api('/api/inventory/output-remap/web-preview', post({from,to,version:data.version,product_code:activeCode,...values()}));
      issues(preview.issues);
      if (!preview.can_confirm) { el('[data-status]').textContent = 'Chưa ghi nhận. Bấm Sửa tại đây ở thông báo để sửa, không cần nhập lại.'; return; }
      const c = preview.changes[0];
      el('[data-preview]').innerHTML = '<section class="card"><h3>Xem trước · Kho chưa thay đổi</h3><p>Chuyển ' + qty(c.qty) + ' ' + esc(c.unit) + ' từ mã bị trừ nhầm sang mặt hàng thực tế đã xuất.</p>' +
        '<table><thead><tr><th>Mặt hàng</th><th>Tồn cuối kỳ sau khi sửa</th></tr></thead><tbody><tr><td>' + esc(c.old_code + ' · ' + c.old_name) + '</td><td>' + qty(c.old_closing_after) + ' ' + esc(c.old_unit) + '</td></tr><tr><td>' + esc(c.new_code + ' · ' + c.new_name) + '</td><td>' + qty(c.new_closing_after) + ' ' + esc(c.unit) + '</td></tr></tbody></table>' +
        '<p>Thông tin trên hóa đơn đã ký giữ nguyên. Thay đổi này chỉ sửa mã hàng trừ kho nội bộ.</p>' +
        '<form data-confirm-form novalidate><label>Người xác nhận<input name="actor" maxlength="100" autocomplete="name"></label>' +
        '<label><input name="confirmed" type="checkbox"> Tôi đã đối chiếu đúng mặt hàng thực tế đã xuất và số lượng sửa.</label>' +
        '<button class="btn btn-primary" data-confirm disabled>Xác nhận sửa và cập nhật kho</button></form></section>';
      el('[data-status]').textContent = 'Kiểm tra tồn sau khi sửa rồi xác nhận ở dưới.';
      el('[data-confirm-form]').onsubmit = confirm;
      field('confirmed').onchange = () => { el('[data-confirm]').disabled = !field('confirmed').checked; };
      el('[data-preview]').scrollIntoView({block:'start'});
    } catch (error) { el('[data-status]').textContent = 'Chưa ghi nhận.'; issues([{field:'reload',message:error.message}]); }
    finally { working(false); }
  }
  async function confirm(event) {
    event.preventDefault(); if (busy || !preview || !field('confirmed').checked) return;
    const actor = field('actor').value.trim();
    if (!actor) { issues([{field:'actor', message:'Điền Người xác nhận trước khi cập nhật kho.'}]); focus('actor'); return; }
    working(true);
    try {
      await api('/api/inventory/output-remap/confirm', post({token:preview.token, actor, confirmed:true}));
      invalidate(); el('[data-editor]').innerHTML = '';
      el('[data-status]').textContent = 'Đã cập nhật kho. Hàng hết âm được bỏ khỏi danh sách; hàng còn âm hiển thị số còn thiếu.';
      await options.onApplied(); await load(false, true);
    } catch (error) { issues([{field:'reload', message:error.message}]); }
    finally { working(false); }
  }
  async function load(preserve = true, afterSave = false) {
    if (busy && !afterSave) return;
    const saved = preserve ? values() : null;
    working(true);
    try {
      const result = await api('/api/inventory/output-remap/shortages?' + new URLSearchParams({from,to}));
      if (!dialog.isConnected) return;
      data = result; renderList(); renderEditor(saved);
      if (!afterSave) el('[data-status]').textContent = data.items.length + ' mặt hàng còn âm. Cập nhật số liệu không thay đổi kho.';
    } catch (error) { issues([{field:/Từ ngày|Đến ngày|ngày đầu|ngày cuối/.test(error.message) ? 'dates' : 'reload',message:error.message}]); }
    finally { working(false); }
  }
  el('[data-close]').onclick = () => { if (!busy) dialog.close(); };
  el('[data-reload]').onclick = () => load(true);
  el('[data-excel]').onclick = () => options.onExcel(() => load(true));
  dialog.addEventListener('input', event => { if (event.target.closest('[data-edit]')) invalidate(); });
  dialog.addEventListener('cancel', event => { if (busy) event.preventDefault(); });
  dialog.addEventListener('close', () => dialog.remove());
  load();
};
